sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "nimble_riot.h"
+#include "net/bluetil/ad.h"
+
+#include "host/ble_hs.h"
+#include "host/util/util.h"
+#include "host/ble_gatt.h"
+#include "services/gap/ble_svc_gap.h"
+#include "services/gatt/ble_svc_gatt.h"
+
+#include "stdio_uart.h"
+#include "periph/uart.h"
+#include "isrpipe.h"
+
+#include "ringbuffer.h"
+#include "stdio_nimble.h"
+
+/* TODO: Why is 261 the absoulte maximum before it stops working? */
+#define MTU_MAX_NIMBLE 261 //BLE_ATT_MTU_MAX
+
+/*
+ * PDU = LL Header (2 bytes)
+ *     + L2CAP Header (4 bytes)
+ *     + Op Code (1 byte)
+ *     + Attribute Handler (2 bytes)
+ *     + MIC (Optional, 4 bytes)
+ *     = 13 bytes
+ * */
+#define PDU_HEADERS_SIZE 13
+#define MAX_MTU_PAYLOAD (MTU_MAX_NIMBLE - PDU_HEADERS_SIZE)
+
+/* ringbuffer for stdin */
+static ringbuffer_t _rb_stdin;
+static char _rb_stdin_mem[CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE];
+
+/* ringbuffer for stdout */
+static ringbuffer_t _rb_stdout;
+static char _rb_stdout_mem[CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE];
+
+static mutex_t _rb_stdout_mutex = MUTEX_INIT;
+static int _current_max_mtu_payload = MAX_MTU_PAYLOAD;
+
+/* intermediate buffer to transfer data between ringbuffer and nimble functions,
+ * which are all based on os_mbuf implementation */
+static char _stdin_read_buf[MAX_MTU_PAYLOAD];
+static char _stdout_write_buf[MAX_MTU_PAYLOAD];
+
+/* information about bluetooth connection */
+static uint16_t _conn_handle;
+static uint16_t _val_handle_stdout;
+static bool _send_stdio = false;
+static bool _sending_in_progress = false;
+static mutex_t _send_mutex = MUTEX_INIT;
+
+static struct ble_gap_event_listener _gap_event_listener;
+
+/**
+ * @brief UUID for stdio service (value: e6d54866-0292-4779-b8f8-c52bbec91e71)
+ */
+static const ble_uuid128_t gatt_svr_svc_stdio_uuid
+        = BLE_UUID128_INIT(0x71, 0x1e, 0xc9, 0xbe, 0x2b, 0xc5, 0xf8, 0xb8, 0x79,
+                           0x47, 0x92, 0x02, 0x66, 0x48, 0xd5, 0xe6);
+
+/**
+ * @brief UUID for stdout characteristic (value: 35f28386-3070-4f3b-ba38-27507e991762)
+ */
+static const ble_uuid128_t gatt_svr_chr_stdout_uuid
+        = BLE_UUID128_INIT(0x62, 0x17, 0x99, 0x7e, 0x50, 0x27, 0x38, 0xba, 0x3b,
+                           0x4f, 0x70, 0x30, 0x86, 0x83, 0xf2, 0x35);
+
+/**
+ * @brief UUID for stdin characteristic (value: ccdd113f-40d5-4d68-86ac-a728dd82f4aa)
+ */
+static const ble_uuid128_t gatt_svr_chr_stdin_uuid
+        = BLE_UUID128_INIT(0xaa, 0xf4, 0x82, 0xdd, 0x28, 0xa7, 0xac, 0x86, 0x68,
+                           0x4d, 0xd5, 0x40, 0x3f, 0x11, 0xdd, 0xcc);
+
+/**
+ * @brief Nimble access callback for stdin characteristic
+ */
+static int gatt_svr_chr_access_stdin(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg);
+
+void debug(const void* buffer, size_t len) {
+    uart_write(STDIO_UART_DEV, (const uint8_t *)buffer, len);
+}
```

Comment:
I don't mind leaving some debug code in, but it should be conditionally enabled. In my branch I had something like:

```

#define PRINTF_BUFSIZ (256 * IS_USED(MODULE_STDIO_NIMBLE_DEBUG))
static char _printf_buf[PRINTF_BUFSIZ];
static int _dbgprintf(const char *format, ...)
{
    if (IS_USED(MODULE_STDIO_NIMBLE_DEBUG)) {
        va_list va;
        va_start(va, format);
        const int ret = vsnprintf(_printf_buf, PRINTF_BUFSIZ, format, va);
        va_end(va);
        uart_write(STDIO_UART_DEV, (const uint8_t *)_printf_buf, ret);
    }
    (void)format;
    (void)_printf_buf;
    return 0;
}
```

Reply:
I added a pseudomodule `stdio_nimble_debug`. `stdio_read` and `stdio_write` still call `uart_write` manually, because those strings they get are not null-terminated.